rust/ql/lib/codeql/rust/elements/internal/StmtListImpl.qll

```diff
@@ -12,17 +11,40 @@ private import codeql.rust.elements.internal.generated.StmtList
  * be referenced directly.
  */
 module Impl {
+  // the following QLdoc is generated: if you need to edit it, do it in the schema file
   /**
-   * A list of statements in a block.
+   * A list of statements in a block, with an optional tail expression at the
+   * end that determines the block's value.
    *
    * For example:
    * ```rust
    * {
    *     let x = 1;
    *     let y = 2;
+   *     x + y
    * }
    * //  ^^^^^^^^^
    * ```
    */
-  class StmtList extends Generated::StmtList { }
+  class StmtList extends Generated::StmtList {
+    /**
+     * Gets the `index`th statement or expression of this statement list (0-based).
+     */
+    AstNode getStmtOrExpr(int index) {
+      result = this.getStatement(index)
+      or
+      index = max(int i | i = -1 or exists(this.getStatement(i))) + 1 and
+      result = this.getTailExpr()
+    }
```

Comment:
As per the above comment I'm not entirely convinced that we want to expose this predicate. But, if we do, then there is in fact already an implementation [here](https://github.com/github/codeql/blob/6678e79239f40b87f482f8e5438fb7676552edb0/rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll#L269). Given that I think we should:
* Delete `getBlockChildNode` and use `getStmtOrExpr` in the control flow graph implementation.
* Use the `getBlockChildNode` implementation for `getStmtOrExpr` as it is a bit simpler.
* Remove the tests added in this PR as the CFG tests will then be exercising `getStmtOrExpr`.

Reply:
That's interesting, I will make some changes...

Reply:
Done the first two.  I could be persuaded to delete the test, but I'm not convinced its redundant - it seems to me that each is testing slightly different areas of code in slightly different ways (there is a lot of overlap).

Reply:
Makes sense! Let's keep the tests.